### PR TITLE
Keep handle on waitset and condition sequences

### PR DESCRIPTION
**Pull request by jacquelinekay**
Connects to ros2/rmw#49

TODO:
- [x] Implement changes for Connext.
- [ ] Improve error handling in rmw_wait so that the error message of `rmw_fini_waitset` gets returned. goto fail might be a better idea than scope exit here.

### CODE REVIEW

**Comment by wjwwood** [comment]
I think that the synchronization should be done at the highest level (rmw and rcl don't know and shouldn't care if they are being used in a multi-threaded way). We should just make sure to document which functions can and cannot be called in a thread-safe way but I don't think the implementations should try to make them thread-safe. Basically I agree with your point above that `rclcpp`/`rclpy` should enforce that the destructor and the trigger function are not called concurrently.

**Comment by dirk-thomas** [comment]
Linter warning: http://ci.ros2.org/job/ci_linux/672/testReport/rmw_opensplice_cpp/cppcheck/error__uninitvar__src_rmw_waitset_cpp_123_/

**Comment by jacquelinekay** [comment]
Should be fixed by 7dc74dd

File: rmw_opensplice_cpp/src/rmw_wait.cpp [hunk 1]
```diff
@@ -45,31 +48,101 @@ rmw_wait(
...
     {
-      // Manually detach conditions, to ensure a clean wait set for next time.
-      DDS::ConditionSeq attached_conditions;
-      waitset.get_conditions(attached_conditions);
-      for (uint32_t i = 0; i < attached_conditions.length(); ++i) {
-        waitset.detach_condition(attached_conditions[i]);
+      // Manually detach conditions and clear sequences, to ensure a clean wait set for next time.
+      if (!waitset) {
+        RMW_SET_ERROR_MSG("waitset handle is null");
+        return;
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: Needs return code.

Same below.

File: rmw_opensplice_cpp/src/rmw_wait.cpp [hunk 2]
```diff
@@ -45,31 +48,101 @@ rmw_wait(
...
+        return )
+      OpenSpliceWaitSetInfo * waitset_info = static_cast<OpenSpliceWaitSetInfo *>(waitset->data);
+      if (!waitset_info) {
+        RMW_SET_ERROR_MSG("WaitSet implementation struct is null");
+        return;
+      }
+
+      DDS::WaitSet * dds_waitset = static_cast<DDS::WaitSet *>(waitset_info->waitset);
+      if (!dds_waitset) {
+        RMW_SET_ERROR_MSG("DDS waitset handle is null");
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: Needs return?

Same below.

File: rmw_opensplice_cpp/src/rmw_wait.cpp [hunk 3]
```diff
@@ -45,31 +48,101 @@ rmw_wait(
...
+      }
+
+      for (uint32_t i = 0; i < attached_conditions->length(); ++i) {
+        bool fixed = false;
+        for (uint32_t j = 0; j < waitset->fixed_guard_conditions->guard_condition_count; ++j) {
+          DDS::GuardCondition * fixed_guard_cond =
+            static_cast<DDS::GuardCondition *>(waitset->fixed_guard_conditions->guard_conditions[j
+            ]);
+          if (fixed_guard_cond == (*attached_conditions)[i]) {
+            fixed = true;
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: This can `break`.

File: rmw_opensplice_cpp/src/rmw_wait.cpp [hunk 4]
```diff
@@ -45,31 +48,101 @@ rmw_wait(
...
+      if (retcode != DDS::RETCODE_OK) {
+        RMW_SET_ERROR_MSG("Failed to get attached conditions for waitset");
+      }
+
+      for (uint32_t i = 0; i < attached_conditions->length(); ++i) {
+        bool fixed = false;
+        for (uint32_t j = 0; j < waitset->fixed_guard_conditions->guard_condition_count; ++j) {
+          DDS::GuardCondition * fixed_guard_cond =
+            static_cast<DDS::GuardCondition *>(waitset->fixed_guard_conditions->guard_conditions[j
+            ]);
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: This wrapping looks weird. Maybe wrap after the open parenthesis?

File: rmw_opensplice_cpp/src/rmw_wait.cpp (hunk 1 above)
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: Nvm this is in the destructor.

File: rmw_opensplice_cpp/src/rmw_waitset.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    if (waitset_info->attached_conditions) {
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+        waitset_info->attached_conditions->~DDS_DCPSUObjSeq(), ConditionSeq)
+      rmw_free(waitset_info->attached_conditions);
+    }
+    if (waitset_info->waitset) {
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(waitset_info->waitset->~WaitSet(), WaitSet)
+      rmw_free(waitset_info->waitset);
+    }
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/rmw_waitset.cpp]: If `waitset_info` is not being free here I would propose to set the freed member variables to zero.

**Comment by dirk-thomas** [comment]
The same comments apply to the Connext PR.

File: rmw_opensplice_cpp/src/rmw_waitset.cpp (hunk 5 above)
Comment (jacquelinekay) [inline rmw_opensplice_cpp/src/rmw_waitset.cpp]: waitset_info is a pointer to waitset_data, which gets freed on the next line.

File: rmw_opensplice_cpp/src/rmw_wait.cpp (hunk 2 above)
Comment (jacquelinekay) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: c490d86

File: rmw_opensplice_cpp/src/rmw_wait.cpp (hunk 3 above)
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: Still applies.

File: rmw_opensplice_cpp/src/rmw_waitset.cpp (hunk 5 above)
Comment (jacquelinekay) [inline rmw_opensplice_cpp/src/rmw_waitset.cpp]: Oh, I didn't interpret your original comment correctly. Should I set all freed member variables to zero/null?

Comment (jacquelinekay) [inline rmw_opensplice_cpp/src/rmw_waitset.cpp]: Fixed.

**Comment by dirk-thomas** [comment]
lgtm.

**Pull request merged**